crml/nft/src/lib.rs

```diff
@@ -974,6 +999,152 @@ decl_module! {
+			let balance = T::MultiCurrency::free_balance(&origin, asset_id);
+			if let Some(balance_after_bid) = balance.checked_sub(amount) {
+				// TODO: review behaviour with 3.0 upgrade: https://github.com/cennznet/cennznet/issues/414
+				// - `amount` is unused
+				// - if there are multiple locks on user asset this could return true inaccurately
+				// - `T::MultiCurrency::reserve(origin, asset_id, amount)` should be checking this internally...
+				let _ = T::MultiCurrency::ensure_can_withdraw(&origin, asset_id, amount, WithdrawReasons::RESERVE, balance_after_bid)?;
+			}
+
+			// try lock funds
+			T::MultiCurrency::reserve(&origin, asset_id, amount)?;
+
+			let mut token_offers = Self::token_offers(token_id);
+			token_offers.push(offer_id);
+			TokenOffers::insert(token_id, token_offers);
+
+			let new_offer = Offer {
+				token_id,
+				asset_id,
+				amount,
+				buyer: origin.clone(),
+				marketplace_id,
+			};
+			Offers::<T>::insert(offer_id, new_offer);
+			NextOfferId::mutate(|i| *i += 1);
+
+			Self::deposit_event(RawEvent::OfferMade(offer_id, amount, asset_id, marketplace_id, origin));
+			Ok(())
+		}
+
+		/// Cancels an offer on a token
+		/// Caller must be the offer buyer
+		#[weight = 0]
+		fn cancel_offer (
+			origin,
+			offer_id: OfferId,
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+			if let Some(offer) = Self::offers(offer_id) {
+				ensure!(offer.buyer == origin, Error::<T>::NotBuyer);
+				T::MultiCurrency::unreserve(&origin, offer.asset_id, offer.amount);
+				Offers::<T>::remove(offer_id);
+				TokenOffers::remove(offer.token_id);
+				Self::deposit_event(RawEvent::OfferCancelled(offer_id));
+				Ok(())
+			} else {
+				Err(Error::<T>::InvalidOffer.into())
+			}
+		}
+
+		/// Accepts an offer on a token
+		/// Caller must be token owner
+		#[weight = 0]
+		fn accept_offer (
+			origin,
+			offer_id: OfferId,
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+			if let Some(offer) = Self::offers(offer_id) {
+				let token_id = offer.token_id;
+				ensure!(Self::token_owner((token_id.0, token_id.1), token_id.2) == origin, Error::<T>::NoPermission);
+
+				let royalties_schedule = Self::check_bundle_royalties(&vec![token_id], offer.marketplace_id)?;
+				let for_royalties = royalties_schedule.calculate_total_entitlement() * offer.amount;
+				let mut for_seller = offer.amount;
+
+				// do royalty payments
+				if !for_royalties.is_zero() {
+					let entitlements = royalties_schedule.entitlements.clone();
+					for (who, entitlement) in entitlements.into_iter() {
+						let royalty = entitlement * offer.amount;
+						let _ = T::MultiCurrency::repatriate_reserved(&offer.buyer, offer.asset_id, &who, royalty)?;
+						for_seller -= royalty;
+					}
+				}
+
+				let seller_balance = T::MultiCurrency::free_balance(&origin, offer.asset_id);
+				let _ = T::MultiCurrency::repatriate_reserved(&offer.buyer, offer.asset_id, &origin, for_seller)?;
```

Comment:
can we find a way to reuse this code block across payments?

Reply:
Yeah was thinking that, will separate out into a separate function